README.md
# Introduction

This is a i18next cache layer to be used in the browser. It will load and cache resources from [MMKV](https://github.com/ammarahm-ed/react-native-mmkv-storage) and can be used in combination with the [chained backend](https://github.com/i18next/i18next-chained-backend).

# Getting started

Source can be loaded via [npm](https://www.npmjs.com/package/i18next-mmkv-storage-cache) or [downloaded](https://github.com/younes200/i18next-mmkv-storage-cache/blob/master/i18nextMMKVStorageCache.min.js) from this repo.

```
# npm package
$ npm install i18next-mmkv-storage-backend
```

Wiring up with the chained backend:

```js
import i18next from 'i18next';
import Backend from 'i18next-chained-backend';
import MMKVStorageBackend from 'i18next-mmkv-storage-backend'; // primary use cache
import XHR from 'i18next-xhr-backend'; // fallback xhr load

i18next
  .use(Backend)
  .init({
    backend: {
      backends: [
        MMKVStorageBackend,  // primary
        XHR                   // fallback
      ],
      backendOptions: [{
        /* below options */
      }, {
        loadPath: '/locales/{{lng}}/{{ns}}.json' // xhr load path for my own fallback
      }]
    }
  });
```

## Cache Backend Options


```js
{
  // prefix for stored languages
  prefix: 'i18next_res_',

  // expiration
  expirationTime: 7*24*60*60*1000,

  // language versions
  versions: {}
};
```

- Contrary to cookies behavior, the cache will respect updates to `expirationTime`. If you set 7 days and later update to 10 days, the cache will persist for 10 days

- Passing in a `versions` object (ex.: `versions: { en: 'v1.2', fr: 'v1.1' }`) will give you control over the cache based on translations version. This setting works along `expirationTime`, so a cached translation will still expire even though the version did not change. You can still set `expirationTime` far into the future to avoid this

Wiring up a service backend with the chained backend:

```js
import i18next from 'i18next';
import Backend from 'i18next-chained-backend';
import MMKVStorageBackend from 'i18next-async-storage-backend'; // primary use cache
import ServiceBackend from 'i18next-service-backend'; // fallback service backend

const TRANSLATION_BACKEND = 'https://api.spacetranslate.com';
const TRANSLATION_BACKEND_PROJECTID = '[projectId]'; // i.e. [projectId].spacetranslate.com

i18next
  .use(Backend)
  .init({
    backend: {
      backends: [
        MMKVStorageBackend,  // primary
        ServiceBackend                   // fallback
      ],
      backendOptions: [{
        // prefix for stored languages
        prefix: 'i18next_res_',

        // expiration
        expirationTime: 7*24*60*60*1000,

        // language versions
        versions: {}
      }, {
        service: TRANSLATION_BACKEND,
        projectId: TRANSLATION_BACKEND_PROJECTID,
        referenceLng: 'en',
        version: 'latest'
      }]
    }
  });
```


CHANGELOG.md
### 1.2.0
- fork to use mmkv

### 1.1.4
- fixes cache save call

### 1.1.1

- initial version


src/index.js
import * as utils from './utils';

// get from whatever version of react native that is being used.
import MMKVStorage from "react-native-mmkv-storage";

function getDefaults() {
  return {
    prefix: 'i18next_res_',
    expirationTime: 7 * 24 * 60 * 60 * 1000,
    versions: {}
  };
}

class Cache {
  constructor(services, options = {}) {
    this.init(services, options);

    this.type = 'backend';
  }

  init(services, options = {}) {
    this.services = services;
    this.options = utils.defaults(options, this.options || {}, getDefaults());
    if(options.instance){
      this.MMKV = options.instance;
    }else{
       this.MMKV = new MMKVStorage.Loader().initialize();
    }

  }

  read(language, namespace, callback) {
    const store = {};
    const nowMS = new Date().getTime();

    if (!this.MMKV) {
      return callback(null, null);
    }

    this.MMKV.getString(`${this.options.prefix}${language}-${namespace}`)
      .then(local => {
        if (local) {
          local = JSON.parse(local);
          if (
            // expiration field is mandatory, and should not be expired
            local.i18nStamp && local.i18nStamp + this.options.expirationTime > nowMS &&

            // there should be no language version set, or if it is, it should match the one in translation
            this.options.versions[language] === local.i18nVersion
          ) {
            delete local.i18nVersion;
            delete local.i18nStamp;
            return callback(null, local);
          }
        }

        callback(null, null);
      })
      .catch(err => {
        console.warn(err);
        callback(null, null);
      });
  }

  save(language, namespace, data) {
    if (this.MMKV) {
      data.i18nStamp = new Date().getTime();

      // language version (if set)
      if (this.options.versions[language]) {
        data.i18nVersion = this.options.versions[language];
      }

      // save
      this.MMKV.setString(`${this.options.prefix}${language}-${namespace}`, JSON.stringify(data));
    }
  }
}

Cache.type = 'backend';

export default Cache;


rollup.config.js
import babel from 'rollup-plugin-babel';
import uglify from 'rollup-plugin-uglify';
import nodeResolve from 'rollup-plugin-node-resolve';
import { argv } from 'yargs';

const format = argv.format || argv.f || 'iife';
const compress = argv.uglify;

const babelOptions = {
  exclude: 'node_modules/**',
  presets: [['es2015', { modules: false }], 'stage-0'],
  babelrc: false
};

const dest = {
  amd: `dist/amd/i18nextMMKVStorageBackend${compress ? '.min' : ''}.js`,
  umd: `dist/umd/i18nextMMKVStorageBackend${compress ? '.min' : ''}.js`,
  iife: `dist/iife/i18nextMMKVStorageBackend${compress ? '.min' : ''}.js`
}[format];

export default {
  entry: 'src/index.js',
  format,
  plugins: [
    babel(babelOptions),
    nodeResolve({ jsnext: true })
  ].concat(compress ? uglify() : []),
  moduleName: 'i18nextMMKVStorageBackend',
  // moduleId: 'i18nextAsyncStorageCache',
  dest
};


i18nextMMKVStorageBackend.js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.i18nextMMKVStorageBackend = factory());
}(this, (function () { 'use strict';

var arr = [];
var each = arr.forEach;
var slice = arr.slice;

function defaults(obj) {
  each.call(slice.call(arguments, 1), function (source) {
    if (source) {
      for (var prop in source) {
        if (obj[prop] === undefined) obj[prop] = source[prop];
      }
    }
  });
  return obj;
}

const ACCESSIBLE = {
  WHEN_UNLOCKED: 'AccessibleWhenUnlocked',
  AFTER_FIRST_UNLOCK: 'AccessibleAfterFirstUnlock',
  ALWAYS: 'AccessibleAlways',
  WHEN_PASSCODE_SET_THIS_DEVICE_ONLY: 'AccessibleWhenPasscodeSetThisDeviceOnly',
  WHEN_UNLOCKED_THIS_DEVICE_ONLY: 'AccessibleWhenUnlockedThisDeviceOnly',
  AFTER_FIRST_UNLOCK_THIS_DEVICE_ONLY: 'AccessibleAfterFirstUnlockThisDeviceOnly',
  ALWAYS_THIS_DEVICE_ONLY: 'AccessibleAlwaysThisDeviceOnly',
};

const MODES = {
  SINGLE_PROCESS: 1,
  MULTI_PROCESS: 2
};

const DATA_TYPES = Object.freeze({
  STRING: 1,
  NUMBER: 2,
  BOOL: 3,
  MAP: 4,
  ARRAY: 5
});

const Loader = require('./src/loader').default;

const MMKVStorage = {
  Loader:Loader,
  MODES: MODES,
  ACCESSIBLE : ACCESSIBLE
};

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

// get from whatever version of react native that is being used.
function getDefaults() {
  return {
    prefix: 'i18next_res_',
    expirationTime: 7 * 24 * 60 * 60 * 1000,
    versions: {}
  };
}

var Cache = function () {
  function Cache(services) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Cache);

    this.init(services, options);

    this.type = 'backend';
  }

  _createClass(Cache, [{
    key: 'init',
    value: function init(services) {
      var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      this.services = services;
      this.options = defaults(options, this.options || {}, getDefaults());
      if (options.instance) {
        this.MMKV = options.instance;
      } else {
        this.MMKV = new MMKVStorage.Loader().initialize();
      }
    }
  }, {
    key: 'read',
    value: function read(language, namespace, callback) {
      var _this = this;

      var store = {};
      var nowMS = new Date().getTime();

      if (!this.MMKV) {
        return callback(null, null);
      }

      this.MMKV.getString('' + this.options.prefix + language + '-' + namespace).then(function (local) {
        if (local) {
          local = JSON.parse(local);
          if (
          // expiration field is mandatory, and should not be expired
          local.i18nStamp && local.i18nStamp + _this.options.expirationTime > nowMS &&

          // there should be no language version set, or if it is, it should match the one in translation
          _this.options.versions[language] === local.i18nVersion) {
            delete local.i18nVersion;
            delete local.i18nStamp;
            return callback(null, local);
          }
        }

        callback(null, null);
      }).catch(function (err) {
        console.warn(err);
        callback(null, null);
      });
    }
  }, {
    key: 'save',
    value: function save(language, namespace, data) {
      if (this.MMKV) {
        data.i18nStamp = new Date().getTime();

        // language version (if set)
        if (this.options.versions[language]) {
          data.i18nVersion = this.options.versions[language];
        }

        // save
        this.MMKV.setString('' + this.options.prefix + language + '-' + namespace, JSON.stringify(data));
      }
    }
  }]);

  return Cache;
}();

Cache.type = 'backend';

return Cache;

})));
